### PR TITLE
feat(ClientOptions): waitGuildTimeout amount client option

### PR DESCRIPTION
**Pull request by kylerchin**
**Please describe the changes this PR makes and why it should be merged:**

Sometimes bots hang while connecting for 15 seconds because 1 single guild is messed up. For bots with Continuous integration, this can create significant impacts to uptime, user experience, and lag.

This adds the optional `fetchGuildTimeout` option under ClientOptions, where the bot developer can optionally specify how long the client should hang for. If not specified, this will remain the default 15 seconds.

I'm happy to discuss how this can be improved, including the names of the options and variables, and I'd also be happy to write documentation!

👉🏻👈🏻 my first pr to this repo, I'm honoured I can be a part of this community.

EDIT: I'm wondering if the "in X sec" should be changed to "in X ms".  

**Status and versioning classification:**

- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

**Comment by kylerchin** [comment]
P.S. I've added this to my own bot, cutting the startup time in half in 1 line of code
Before
Before
image
After
After

File: src/client/websocket/WebSocketShard.js
Comment (DTrombett) [inline src/client/websocket/WebSocketShard.js:487]: 🧹
```suggestion
    const guildTimeoutAmount = this.manager.client.options.fetchGuildTimeout || 1500;
```

File: src/client/websocket/WebSocketShard.js
Comment (DTrombett) [inline src/client/websocket/WebSocketShard.js:493]:
```suggestion
            `${hasGuildsIntent ? `in ${Math.round(guildTimeoutAmount / 1000)} seconds` : ''}.\n   Unavailable guild count: ${this.expectedGuilds.size}`,
```

**Review by DTrombett — COMMENTED** [review]
This is also missing documentation and validation in the Client class.

**Comment by thehackerboi69github** [comment]
Wouldn't `shardReadyTimeout` be a more appropriate name for this?

File: src/client/websocket/WebSocketShard.js
Comment (ckohen) [reply src/client/websocket/WebSocketShard.js:487]: You sure you want 0 to be treated as default? If not use ??

Comment (NotSugden) [reply src/client/websocket/WebSocketShard.js:487]: what about setting the default here
https://github.com/discordjs/discord.js/blob/d16ada97087e0324134484c9c0724b52a2839bdf/src/util/Options.js#L110-L146

Comment (ckohen) [reply src/client/websocket/WebSocketShard.js:487]: I mean yes that would be preferred once validation is added 😎

Comment (DTrombett) [reply src/client/websocket/WebSocketShard.js:487]: > You sure you want 0 to be treated as default? If not use ??

If `15000` is added as a default option this isn't needed as `this.manager.client.options.fetchGuildTimeout` can be used... However, why you should use 0? Does this mean no timeout at all or no guilds at all?

File: src/client/websocket/WebSocketShard.js
Comment (kylerchin) [reply src/client/websocket/WebSocketShard.js:487]: What if we just used ms?

**Comment by kylerchin** [comment]
I've tried my best to make the recommended changes, let me know if we need to change anything else or we have any other thoughts about this before merging.

**Comment by ImRodry** [comment]
Yeah I’m pretty sure that text you added goes over the 120 char limit, you should install ESLint to be aware of these things since the CI requires approval to run

File: src/client/websocket/WebSocketShard.js
Comment (thehackerboi69github) [inline src/client/websocket/WebSocketShard.js:488]:
```suggestion
            `${hasGuildsIntent ? ` in ${guildTimeoutAmount} milliseconds` : ''}.\n   Unavailable guild count: ${this.expectedGuilds.size}`,
```

Comment (ImRodry) [reply src/client/websocket/WebSocketShard.js:488]: Should probably also say ms instead of milliseconds

Comment (kylerchin) [reply src/client/websocket/WebSocketShard.js]: > Yeah I’m pretty sure that text you added goes over the 120 char limit, you should install ESLint to be aware of these things since the CI requires approval to run

thank you! I've corrected this issue now and it passed the lint :) @ImRodry 
I've learned a lot and I hope this PR is helpful to all discord js bot devs!

File: src/client/websocket/WebSocketShard.js
Comment (ImRodry) [inline src/client/websocket/WebSocketShard.js:490]: I think this isn't allowed either. Just format with prettier and it will adjust it for you

Comment (kylerchin) [reply src/client/websocket/WebSocketShard.js:490]: Thank you, i will try this in a little bit.

File: src/client/websocket/WebSocketShard.js
Comment (DTrombett) [reply src/client/websocket/WebSocketShard.js]: Well, it's the same. I personally think that if the option is in ms maybe the message logged should be in ms too but it's better that official djs collaborators decide lol

Comment (ImRodry) [reply src/client/websocket/WebSocketShard.js]: The option should probably be in seconds instead since you're not gonna want nor need a timeout of less than 1 second

**Comment by DTrombett** [comment]
You can add the default value for this option in the `Options.createDefault()` static method:
https://github.com/discordjs/discord.js/blob/stable/src/util/Options.js#L102
And add validation in the `Client#_validateOptions()` method:
https://github.com/discordjs/discord.js/blob/stable/src/client/Client.js#L542

**Comment by kylerchin** [comment]
I'm debating if this should be titled "fetchGuildTimeout" or "fetchGuildsTimeout" or something else. I'm happy to discuss this and improve the experience for developers.

File: src/client/websocket/WebSocketShard.js
Comment (kylerchin) [reply src/client/websocket/WebSocketShard.js]: I've used the prettier version! it looks so much better :) thank you

File: src/client/websocket/WebSocketShard.js
Comment (kylerchin) [reply src/client/websocket/WebSocketShard.js]: I think it's kinda weird to have decimals in the options but I'm not entirely sure. I think the console log and options should be the same formatting (in milliseconds) to make the experience smoother for devs.

Comment (ImRodry) [reply src/client/websocket/WebSocketShard.js]: But you shouldn’t need decimals for the same reason you shouldn’t need milliseconds. Your guilds will never arrive that fast so it’s unreasonable to set a lower timeout. If the dev REALLY wants that then they can use decimals, but I don’t see why they would

File: src/client/websocket/WebSocketShard.js
Comment (monbrey) [inline src/client/websocket/WebSocketShard.js:483]:
```suggestion
    const guildTimeoutAmount = this.manager.client.options.fetchGuildTimeout || 15000;
```

File: src/client/websocket/WebSocketShard.js
Comment (kylerchin) [reply src/client/websocket/WebSocketShard.js]: Thank you for pointing that out. I was looking at the client options and the request-based ones all use milliseconds. Devs are already quite familiar with setTimeout in JS and other similar stuff being based on milliseconds.

Does this change have to be fulfilled for the option to be merged into the repo?

File: src/util/Options.js
Comment (thehackerboi69github) [inline src/util/Options.js:74]:
```suggestion
 * @property {number} [fetchGuildTimeout=15000] Time in milliseconds to wait for missing guilds to be received before marking as ready (0 to skip this timeout)
```

File: src/client/websocket/WebSocketShard.js
Comment (ImRodry) [reply src/client/websocket/WebSocketShard.js]: Yeah that’s true. I’m not a maintainer I’m just giving my personal opinion so let’s wait for one of the @discordjs/the-big-3 to say something about this

File: src/util/Options.js
Comment (ImRodry) [reply src/util/Options.js]: The only good thing about this change is that it mentions that 0 skips the timeout. Other than that, why??
You removed "Clients with the GUILDS intent" and "If not specified, the default is 15 seconds" which imo are quite important mentions, what's the purpose of this?

File: src/client/websocket/WebSocketShard.js
Comment (vladfrangu) [inline src/client/websocket/WebSocketShard.js:479]:
```suggestion
    // * The timeout is 15 seconds by default
```

File: src/client/websocket/WebSocketShard.js
Comment (vladfrangu) [inline src/client/websocket/WebSocketShard.js:480]:
```suggestion
    // * This can be optionally be changed in the client options via the `fetchGuildTimeout` option
```

File: src/client/websocket/WebSocketShard.js
Comment (vladfrangu) [inline src/client/websocket/WebSocketShard.js:481]: While this is documentation in internal code, users shouldn't use a timeout of 0 without acknowledging the difference it can make

File: src/client/websocket/WebSocketShard.js
Comment (vladfrangu) [reply src/client/websocket/WebSocketShard.js:479]: Also this default should be in the client option, so no need for `?? 15_000`

File: src/util/Options.js
Comment (vladfrangu) [inline src/util/Options.js:131]:
```suggestion
      fetchGuildTimeout: 15_000,
```

File: src/util/Options.js
Comment (vladfrangu) [inline src/util/Options.js:72]:
```suggestion
 * @property {number} [fetchGuildTimeout=15000] Time in milliseconds that clients with the `GUILDS` intent should hang
```

**Review by vladfrangu — DISMISSED** [review]

File: src/util/Options.js
Comment (thehackerboi69github) [inline src/util/Options.js:73]:
```suggestion
 * and wait for missing guilds to be received before starting the bot
```

File: src/client/websocket/WebSocketShard.js
Comment (thehackerboi69github) [inline src/client/websocket/WebSocketShard.js:483]:
```suggestion
    const guildTimeoutAmount = this.manager.client.options.fetchGuildTimeout;
```

Comment (thehackerboi69github) [reply src/client/websocket/WebSocketShard.js:483]: Defaults will be assigned in the client constructor when not specified, so there's no need to specify it here as well

**Review by thehackerboi69github — CHANGES_REQUESTED** [review]

File: src/util/Options.js
Comment (thehackerboi69github) [inline src/util/Options.js:74]: Is this line really needed? The `[fetchGuildTimeout=15000]` in L72 will already show the default in the docs.

File: src/util/Options.js
Comment (1chiSensei) [inline src/util/Options.js:72]:
```suggestion
 * @property {number} [fetchGuildTimeout=15000] Time in milliseconds that bots with the `GUILDS` intent should hang
```

Shouldn't be this called bots instead of clients?

**Review by 1chiSensei — CHANGES_REQUESTED** [review]

File: src/util/Options.js
Comment (ImRodry) [reply src/util/Options.js]: Not at all

File: src/util/Options.js
Comment (SpaceEEC) [inline src/util/Options.js:75]: A few non-binding rephrasing suggestions:
- `hang and wait`: seems redundant
- `missing`: feels a bit misleading (might just be me)
- `starting the bot`: `Client#login` starts it. It's just not ready yet, a small nit.
- `skip this timeout`: `not wait` seems to be a bit more concise

```suggestion
 * @property {number} [waitGuildTimeout=15000] Time in milliseconds that bots with the `GUILDS` intent should wait for
 * initial guild's data to be received before emitting ready. (0 to not wait)
```

**Review by SpaceEEC — COMMENTED** [review]
Maybe rename the option to `waitGuildTimeout` or something the like that?
I think that `fetch` might be a bit misleading as this isn't something we directly "requested" in that sense, that's just part of the login and ready flow. (Probably not a good explanation, but I can't come up with anything better at the moment)

File: src/client/websocket/WebSocketShard.js
Comment (thehackerboi69github) [inline src/client/websocket/WebSocketShard.js]:
```suggestion
    const { waitGuildTimeout } = this.manager.client.options;
```

File: src/client/websocket/WebSocketShard.js
Comment (thehackerboi69github) [inline src/client/websocket/WebSocketShard.js:489]:
```suggestion
            `${hasGuildsIntent ? ` in ${waitGuildTimeout} ms` : ''}.\nUnavailable guild count: ${
```

File: src/client/websocket/WebSocketShard.js
Comment (thehackerboi69github) [inline src/client/websocket/WebSocketShard.js:500]:
```suggestion
      hasGuildsIntent ? waitGuildTimeout : 0,
```

File: typings/index.d.ts
Comment (thehackerboi69github) [inline typings/index.d.ts]:
```suggestion
  waitGuildTimeout?: number;
```

**Comment by kylerchin** [comment]
@SpaceEEC the changes have been applied

**Review by thehackerboi69github — CHANGES_REQUESTED** [review]

File: src/util/Options.js
Comment (SpaceEEC) [inline src/util/Options.js:72]: To comply with the line length limit:
```suggestion
 * @property {number} [waitGuildTimeout=15000] Time in milliseconds that Clients with the GUILDS intent should wait for
 * missing guilds to be recieved before starting the bot. If not specified, the default is 15 seconds.
```

File: src/util/Options.js
Comment (SpaceEEC) [inline src/util/Options.js:120]: I don't think we want to revert those:
```suggestion
      restWsBridgeTimeout: 5_000,
      restRequestTimeout: 15_000,
```

**Review by SpaceEEC — CHANGES_REQUESTED** [review]
You also need to add the documented default for `waitGuildTimeout` in `Options.createDefault`, otherwise an error will be thrown if omitted.

**Review by SpaceEEC — APPROVED** [review]

File: src/client/websocket/WebSocketShard.js
Comment (vladfrangu) [inline src/client/websocket/WebSocketShard.js:481]: Setting this to 0 *might* break many things. Are we sure we should allow it?

**Review by vladfrangu — CHANGES_REQUESTED** [review]

**Review by kyranet — APPROVED** [review]

**Comment by kylerchin** [comment]
everything in here looks done! 

let me know if anything else needs to be changed, otherwise, this will be it! happy that this will get merged soon!

**Comment by thehackerboi69github** [comment]
The default value is still not in  `Options.createDefault()`

**Comment by Jiralite** [comment]
Check the reviews I submitted also.

File: src/util/Options.js
Comment (almeidx) [inline src/util/Options.js:73]:
```suggestion
 * @property {number} [waitGuildTimeout=15_000] Time in milliseconds that Clients with the GUILDS intent should wait for
 * missing guilds to be received before starting the bot.
```

**Review by almeidx — CHANGES_REQUESTED** [review]

File: src/util/Options.js
Comment (almeidx) [reply src/util/Options.js]: And the
> If not specified, the default is 15 seconds.

bit should be removed

**Comment by thehackerboi69github** [comment]
What does.that last commit have to do with this PR?

**Comment by kylerchin** [comment]
It fixes a bug in the discord name acronym system. Apologies.

File: src/structures/BaseGuild.js
Comment (kyranet) [inline src/structures/BaseGuild.js:71]: Please revert this, do not introduce unrelated changes in a PR.

**Review by kyranet — CHANGES_REQUESTED** [review]

**Comment by kylerchin** [comment]
I reverted the changes. Really sorry.

**Comment by kylerchin** [comment]
Those changes have been moved to https://github.com/discordjs/discord.js/pull/7050/

**Review by kyranet — APPROVED** [review]

**Comment by thehackerboi69github** [comment]
> The default value is still not in  `Options.createDefault()`

**Review by thebigwolfy — APPROVED** [review]

**Review by vladfrangu — CHANGES_REQUESTED** [review]
The default is still not added in `Options.createDefault()`

**Comment by kylerchin** [comment]
Apologies for not updating this, I will change it.

Sorry, i've been busy with finals.

**Comment by kylerchin** [comment]
Again, i'm really sorry about the delay. I added it to options now

**Comment by vladfrangu** [comment]
All I want for Christmas..issss... a rebaseeeeee 🎵

**Review by iCrawl — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

**Pull request merged**
